Editing files when using su is not possible

If you try to edit a file that doesnt belong to you, like for example a file from your nix config you will get the following error:

```
❯ sudo opsops edit inplan.yaml
📝 Opening file for editing: inplan.yaml
⨯ Failed to get Age key: 1Password CLI returned an error: [ERROR] 2025/06/30 10:57:00 Can't continue. We can't safely access "/Users/jhofer/.config/op" because it's not owned by the current user. Change the owner or logged in user and try again.
```

--- a/.sops.yaml
+++ b/.sops.yaml
@@ -1,2 +1,7 @@
-creation_rules: []
+creation_rules:
+- path_regex: .*\.(yaml|yml|json|env|ini|sops|txt)$
+  age: null
+  key_groups:
+  - age:
+    - age1kula8luezle3jds5mergwvtpmyu5s6tqnt3dzfxeuukuacq9xeds8lvfe7
 onepassworditem: op://Personal/test/Private Key

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,6 +15,7 @@ serde = { version = "1.0.219", features = ["derive"]}
 serde_json = "1.0.140"
 serde_yaml = "0.9.34"
 tempfile = "3.20.0"
+users = "0.11.0"
 which = "8.0.0"
 
 [build-dependencies]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "opsops"
-version = "1.3.0"
+version = "1.3.1"
 edition = "2024"
 
 [dependencies]

--- a/justfile
+++ b/justfile
@@ -42,7 +42,7 @@ test-one TEST:
 
 # Run clippy lints
 lint:
-    @cargo clippy -- -D warnings
+    @cargo clippy
 
 # Format code
 fmt:

--- a/src/commands/doctor.rs
+++ b/src/commands/doctor.rs
@@ -2,13 +2,62 @@ use crate::{
     GlobalContext,
     util::{
         op_key::{extract_public_key, get_age_key_from_1password},
-        print_status::{print_error, print_info, print_success, print_warning},
+        print_status::{print_error, print_success, print_warning},
         sops_config::read_or_create_config,
     },
 };
 use colored::Colorize;
 
 pub fn doctor(context: &GlobalContext) {
+    match which::which("sops") {
+        Ok(path) => {
+            let version = std::process::Command::new(&path)
+                .arg("--version")
+                .output()
+                .ok()
+                .and_then(|o| String::from_utf8(o.stdout).ok())
+                .map(|out| out.lines().next().unwrap_or("unknown").to_string())
+                .unwrap_or_else(|| "unknown".to_string());
+            print_success(format!(
+                "{} {} {}",
+                "Found sops:".green(),
+                path.display(),
+                version.trim().dimmed()
+            ));
+        }
+        Err(_) => {
+            print_error(format!(
+                "{}",
+                "sops is not installed or not found in PATH. Please install sops.".red()
+            ));
+            return;
+        }
+    }
+    match which::which("op") {
+        Ok(path) => {
+            let version = std::process::Command::new(&path)
+                .arg("--version")
+                .output()
+                .ok()
+                .and_then(|o| String::from_utf8(o.stdout).ok())
+                .unwrap_or_else(|| "unknown".to_string());
+            print_success(format!(
+                "{} {} {}",
+                "Found 1Password CLI (op):".green(),
+                path.display(),
+                version.trim().dimmed()
+            ));
+        }
+        Err(_) => {
+            print_error(format!(
+                "{}",
+                "1Password CLI (op) is not installed or not found in PATH. Please install op."
+                    .red()
+            ));
+            return;
+        }
+    }
+
     let config = match read_or_create_config(context) {
         Ok(c) => c,
         Err(err) => {
@@ -24,9 +73,9 @@ pub fn doctor(context: &GlobalContext) {
         ));
         return;
     } else {
-        print_info(format!(
+        print_success(format!(
             "{} {}\n",
-            "1Password reference found in .sops.yaml:".green(),
+            "1P item found in .sops.yaml:".green(),
             config.onepassworditem
         ));
     }
@@ -81,7 +130,7 @@ pub fn doctor(context: &GlobalContext) {
                     if derived_public_key == *key {
                         print_success(format!(
                             "{} {}",
-                            "Found matching public key in key group:".green(),
+                            "Found matching public key in .sops.yaml:".green(),
                             key
                         ));
                         found = true;

--- a/src/commands/doctor.rs
+++ b/src/commands/doctor.rs
@@ -75,7 +75,7 @@ pub fn doctor(context: &GlobalContext) {
     } else {
         print_success(format!(
             "{} {}\n",
-            "1P item found in .sops.yaml:".green(),
+            "1Password item found in .sops.yaml:".green(),
             config.onepassworditem
         ));
     }

--- a/src/util/op.rs
+++ b/src/util/op.rs
@@ -2,7 +2,7 @@ use serde::Deserialize;
 use std::process::Command;
 use users::os::unix::UserExt;
 
-use crate::util::print_status::print_info;
+use crate::util::print_status::{print_info, print_warning};
 
 use super::print_status::print_error;
 
@@ -98,9 +98,15 @@ pub fn op_command() -> Command {
                 // Set HOME to the user's home directory
                 if let Some(home) = user.home_dir().to_str() {
                     cmd.env("HOME", home);
+                } else {
+                    print_warning("Couldn't get home directory of sudo user");
                 }
                 return cmd;
+            } else {
+                print_warning("Couldn't get sudo user by name");
             }
+        } else {
+            print_warning("Sudo User is empty!");
         }
     }
     Command::new("op")

--- a/src/util/op.rs
+++ b/src/util/op.rs
@@ -106,7 +106,7 @@ pub fn op_command() -> Command {
                 print_warning("Couldn't get sudo user by name");
             }
         } else {
-            print_warning("Sudo User is empty!");
+            print_warning("Environment variable SUDO_USER is set but empty");
         }
     }
     Command::new("op")

--- a/src/util/op.rs
+++ b/src/util/op.rs
@@ -1,5 +1,8 @@
 use serde::Deserialize;
 use std::process::Command;
+use users::os::unix::UserExt;
+
+use crate::util::print_status::print_info;
 
 use super::print_status::print_error;
 
@@ -80,8 +83,31 @@ pub struct OpItem {
     pub(crate) fields: Vec<OpItemField>,
 }
 
+/// Helper to run the `op` CLI as the invoking user if running under sudo.
+pub fn op_command() -> Command {
+    use std::env;
+    use std::os::unix::process::CommandExt;
+
+    if let Ok(sudo_user) = env::var("SUDO_USER") {
+        if !sudo_user.is_empty() {
+            // Get the user's UID and GID
+            if let Some(user) = users::get_user_by_name(&sudo_user) {
+                let mut cmd = Command::new("op");
+                cmd.uid(user.uid());
+                cmd.gid(user.primary_group_id());
+                // Set HOME to the user's home directory
+                if let Some(home) = user.home_dir().to_str() {
+                    cmd.env("HOME", home);
+                }
+                return cmd;
+            }
+        }
+    }
+    Command::new("op")
+}
+
 pub fn op_item_create(item: OpItem) {
-    let mut cmd = Command::new("op");
+    let mut cmd = op_command();
 
     cmd.arg("item")
         .arg("create")
@@ -90,7 +116,7 @@ pub fn op_item_create(item: OpItem) {
         .arg("--title")
         .arg(&item.title)
         .arg("--category")
-        .arg(item.category.as_str()); // implement as_str on OpCategory
+        .arg(item.category.as_str());
 
     for field in item.fields {
         let field_str = match (&field.section, &field.field_type) {
@@ -118,7 +144,7 @@ pub fn op_item_create(item: OpItem) {
 }
 
 pub fn _op_item_get(item_name: &str, field: &str) -> Option<String> {
-    let output = Command::new("op")
+    let output = op_command()
         .arg("item")
         .arg("get")
         .arg(item_name)
@@ -139,8 +165,7 @@ pub fn _op_item_get(item_name: &str, field: &str) -> Option<String> {
 }
 
 pub fn get_vaults() -> Option<Vec<String>> {
-    // Execute the `op vault list --format=json` command
-    let output_json = Command::new("op")
+    let output_json = op_command()
         .arg("vault")
         .arg("list")
         .arg("--format=json")
@@ -148,7 +173,6 @@ pub fn get_vaults() -> Option<Vec<String>> {
         .ok()?;
 
     if output_json.status.success() {
-        // Parse the JSON response
         let vaults: Vec<Vault> = match serde_json::from_slice(&output_json.stdout) {
             Ok(v) => v,
             Err(e) => {
@@ -157,7 +181,6 @@ pub fn get_vaults() -> Option<Vec<String>> {
             }
         };
 
-        // Extract vault names
         let vault_names: Vec<String> = vaults.into_iter().map(|vault| vault.name).collect();
         Some(vault_names)
     } else {
@@ -170,8 +193,7 @@ pub fn get_vaults() -> Option<Vec<String>> {
 }
 
 pub fn get_items(vault: &String) -> Option<Vec<String>> {
-    // Execute the `op vault list --format=json` command
-    let output_json = Command::new("op")
+    let output_json = op_command()
         .arg("item")
         .arg("list")
         .arg("--vault")
@@ -181,7 +203,6 @@ pub fn get_items(vault: &String) -> Option<Vec<String>> {
         .ok()?;
 
     if output_json.status.success() {
-        // Parse the JSON response
         let vaults: Vec<ListItem> = match serde_json::from_slice(&output_json.stdout) {
             Ok(v) => v,
             Err(e) => {
@@ -190,7 +211,6 @@ pub fn get_items(vault: &String) -> Option<Vec<String>> {
             }
         };
 
-        // Extract vault names
         let item_names: Vec<String> = vaults.into_iter().map(|item| item.title).collect();
         Some(item_names)
     } else {
@@ -203,8 +223,7 @@ pub fn get_items(vault: &String) -> Option<Vec<String>> {
 }
 
 pub fn get_fields(item: &String, vault: &String) -> Option<Vec<String>> {
-    // Execute the `op vault list --format=json` command
-    let output_json = Command::new("op")
+    let output_json = op_command()
         .arg("item")
         .arg("get")
         .arg(item)
@@ -215,7 +234,6 @@ pub fn get_fields(item: &String, vault: &String) -> Option<Vec<String>> {
         .ok()?;
 
     if output_json.status.success() {
-        // Parse the JSON response
         let fields: ItemFields = match serde_json::from_slice(&output_json.stdout) {
             Ok(v) => v,
             Err(e) => {
@@ -224,7 +242,6 @@ pub fn get_fields(item: &String, vault: &String) -> Option<Vec<String>> {
             }
         };
 
-        // Extract vault names
         let item_names: Vec<String> = fields.fields.into_iter().map(|item| item.label).collect();
         Some(item_names)
     } else {

--- a/src/util/op.rs
+++ b/src/util/op.rs
@@ -2,7 +2,7 @@ use serde::Deserialize;
 use std::process::Command;
 use users::os::unix::UserExt;
 
-use crate::util::print_status::{print_info, print_warning};
+use crate::util::print_status::print_warning;
 
 use super::print_status::print_error;
 

--- a/src/util/op_key.rs
+++ b/src/util/op_key.rs
@@ -1,4 +1,7 @@
-use crate::{GlobalContext, util::sops_config::read_or_create_config};
+use crate::{
+    GlobalContext,
+    util::{op::op_command, sops_config::read_or_create_config},
+};
 use age::{
     secrecy::{ExposeSecret, SecretString},
     x25519::Identity,
@@ -30,15 +33,9 @@ pub fn get_age_key_from_1password(context: &GlobalContext) -> Result<String, Str
         config.onepassworditem
     };
 
-    // print_info(format!(
-    //     "{} {}",
-    //     "🔑 Retrieving Age key from".dimmed(),
-    //     op_reference.dimmed()
-    // ));
-
     // Run the op command to get the key
     // Format: op://<vault>/<item>/<field>
-    let output = Command::new("op")
+    let output = op_command()
         .arg("read")
         .arg(&op_reference)
         .output()

--- a/src/util/op_key.rs
+++ b/src/util/op_key.rs
@@ -7,7 +7,7 @@ use age::{
     x25519::Identity,
 };
 use colored::Colorize;
-use std::{process::Command, str::FromStr};
+use std::str::FromStr;
 
 use super::print_status::print_error;
 

--- a/src/util/sops_command.rs
+++ b/src/util/sops_command.rs
@@ -121,6 +121,10 @@ mod tests {
 
     #[test]
     fn test_builder_runs_valid_command() {
+        if which::which("sops").is_err() {
+            eprintln!("Skipping test_builder_runs_valid_command: 'sops' binary not found in PATH.");
+            return;
+        }
         let context = mock_context(None);
 
         let output = SopsCommandBuilder::new(&context)
@@ -139,6 +143,10 @@ mod tests {
 
     #[test]
     fn test_env_injection() {
+        if which::which("sops").is_err() {
+            eprintln!("Skipping test_env_injection: 'sops' binary not found in PATH.");
+            return;
+        }
         let context = mock_context(Some(
             "AGE-SECRET-KEY-1AM036DUJQ8RTJ84N7JTJECSV6FXFM3DCM9F4VEX4ZPL4M3VDA6FQLVJSUR"
                 .to_string(),

